### PR TITLE
Refactor and Fix Jalali issues

### PR DESCRIPTION
**Pull request by alitaheri**
I have removed some formatting functions as they were unnecessary and replaced them with `format`. Also, I have updated my custom utils and updated the examples

### CODE REVIEW

**Comment by dmtrKovalenko** [comment]
Oh, I apologize for inconvenience, but in the replace-moment PR we have discussed and realized that it would be more consistent for users to let them override any displayed strings via extending utils, so may you revert back all displayed methods. 
I`m planning to create separate page about customization displayed string in the docs. 
Thank you! :)

**Comment by alitaheri** [comment]
@dmtrKovalenko I couldn't find the discussion 😅 Can you give me a link?

Also, I don't think it's a god idea to have both a format function and some format specific functions. this duality can lead to serious bugs. As you said yourself, the implementors can parse the format string and act accordingly as I did [here](https://github.com/alitaheri/material-ui-pickers-jalali-utils/blob/master/index.js#L78). But having the both is a maintenance headache. We should either remove the `format` method entirely, or solely use it for formatting. I am strongly against having both ways. we should pick one of the two.

cc @cherniavskii

**Comment by cherniavskii** [comment]
@alitaheri I think this https://github.com/dmtrKovalenko/material-ui-pickers/pull/218#issuecomment-365604881 is a discussion @dmtrKovalenko is talking about. From that point of view, it makes sense.

That being said, I'd rather prefer single format function, since it's less code to maintain and in that case date utils library is responsible for those functions.
It's a bit harder to override displayed strings in this case, but it's still possible.

**Comment by dmtrKovalenko** [comment]
@alitaheri  I think mapping strings is not really good idea. Because if somebody will need to defenetly override something it would make a "bad code" something like 
```js
if (formatString === 'mm')
  return date.format('dd')
```
It`s also bad from the maintenance prospective, because it would be hard to know that additional using of format would not break anything.

Now just a format prop is using to display custom formatted dates in the textfield and we should leave it. And for all other displayed dates (in headers and so on) we should create a method with cusom formatting. But I think it would be more convenient to have an ability globally override displayed strings in pickers if it needs for someone.

**Comment by alitaheri** [comment]
@cherniavskii At first I was skeptical, but after giving it some thought, I realized it was the ideal solution to replace all formatting functions with a unified function. It's future proof.

**Comment by dmtrKovalenko** [comment]
@alitaheri 
Also maybe you should use class syntax to extend the prop, instead of reimplementing functions? 
It would be better to just inherit not changing functions by the jalali-utils?

**Comment by alitaheri** [comment]
@dmtrKovalenko Almost all date utilities understand the standard formatting, besides, it's really rare for someone to want to change the current behavior of the date picker. also, if someone is worried about future changes breaking their code, they can use regexp matching.

It is also very rare for someone to want to change the format string. The reason I did it was because of the limitations of the library I was using, the format string is a global standard and almost every library understands it. I believe we won't see another person trying to change the format string.

> Also maybe you should use class syntax to extend the utils

What good would that do? the utils class isn't instantiated anywhere, and extending a class does not inherit the static methods. It would only pollute the prototype chain.

**Comment by dmtrKovalenko** [comment]
@alitaheri 
We would need to create them not static to make localization works for date-fns. 😭

**Comment by alitaheri** [comment]
@dmtrKovalenko It doesn't matter at all you know. the utils should only have some methods on it. doesn't matter if it's an instance, plain object, function, or class constructor. Implementation has nothing to do with the API. as long as it works. and this is working just fine. I would suggest putting our ideas to vote and see what the community thinks. But in the end it's your call, on both formatting methods and class syntax.

P.S. If I want to use the class syntax I'll need a transpiler too, I wanted to keep things simple, so I just used es5.

**Comment by dmtrKovalenko** [comment]
@alexander-morozov 
I have considered inheritance as a way to let users just **override** needed functionality. It may be used for localization or some other purpose. You always can do inheritance based on prototypes, but thus you will not need to implement for example `setDays` and other methods.

**Comment by alitaheri** [comment]
@dmtrKovalenko ok I can change it into a class.

**Comment by cherniavskii** [comment]
@dmtrKovalenko is class the only way to override some methods? Isn't it possible with simple object? Adding transpiler to jaali utils sounds like a real overkill

**Comment by dmtrKovalenko** [comment]
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Inheritance_and_the_prototype_chain
It can be done easily with native javascript prototype

**Comment by cherniavskii** [comment]
@alitaheri maybe adding `material-ui-pickers` to `peerDependencies` will solve the problem?

**Comment by dmtrKovalenko** [comment]
@alitaheri 
You can add pickers to the peer dependency (And you should do that, because jalali-utils are useless without pickers)
Then you will need to import jalali utils relatively ('/utils/moment-utils') extend current pickers and just override the methods you need to override all functions instead of the same like these
```js 
utils.isAfterDay = function isAfterDay(date, value) {
  return date.isAfter(value, 'day');
}

utils.isBeforeDay = function isBeforeDay(date, value) {
  return date.isBefore(value, 'day');
}
```

static methods can be overriden like
```
import utils from '...'
utils.getDisplayedHeaderText = ...
```

But firstly we need to redo current functionality to instance classes. Because we need to integrate localization for date-fns, and its just only one way to perform it, because we need to pass locale string to each function.

**Comment by alitaheri** [comment]
I'm really uncomfortable with these ideas. This really isn't a complex matter.

### About the formatting functions:

Almost every library understands the standard formats. And:

> The most widespread civil calendar and de facto international standard is the Gregorian calendar. Though that calendar is associated with the Catholic Church and the papacy, it has been adopted, as a matter of convenience, by many secular and non-Christian countries although some countries use other calendars.

From [wikipedia](https://en.wikipedia.org/wiki/Civil_calendar).

I think it's over-engineering what I previously did with the formatting functions. I regret it, a single format methods is way better in terms of maintenance. The utils implementors should simply write future proof [matching regexps](https://regex101.com/r/ijowTo/1).

Imaging having to update 10 utils library simply because we want to add a new display string. But with a single format function it may never come to that. Besides, all the utils will become useless because that missing function will definitely break a lot of code (ReferenceError). with single format function, however, it might **rarely** display wrong value. Only in case the utils library was implemented poorly.

A single format function is enough really, all those useless functions are really unnecessary. 

### About depending on the main library:

Since it's a direct dependency, (That means this library **will not work** without relying on the main lib) I cannot add it as a peer dependency. that will defeat the purpose.

Peer Dependencies mean: this library is useless without it's peer. Not that it won't work without it.

besides, aren't we going to put the utils in different repositories eventually?

Another point is dependency graph. If I rely both on moment and moment-utils, their version should always match, otherwise we'll end up with 2 different installation with will break localization. Remember? moment defines locale globally too, so that's alarming!

I'm strongly against, relying on anything other than moment-jalaali. It's not a good idea to rely on another library only to bring a few functions over. It's nothing more than few lines of code. but the overhead of maintaining dependency graph is exhausting. I can refer you to the troubles react libraries have with react dependency when a major version is released. moment is like react, there shouldn't be more that one installation, or things will break.

**Comment by dmtrKovalenko** [comment]
@alitaheri I agree with you in the case of format implementation, but what if somebody need to completely override displayed format in the picker. 
Example - somebody needs to display days with leading zeros? The code he would need to write

```js
if (formatString === 'd') {
  return format(date, 'DD')
}
```

This looks awful) Maybe then we need to integrate custom formatting strings passed with for each formatting value

**Comment by alitaheri** [comment]
@dmtrKovalenko I see your point 👍 👍 👍 alright.

I'll keep the display functions, but I still better not inherit from the moment-utils. because if I do, `moment-jalaali` and `material-ui-pickers` should **always** have compatible `moment` dependency. Is that alright?

**Comment by alitaheri** [comment]
@dmtrKovalenko @cherniavskii This is ready to merge. please have a look.

File: docs/src/Routes/Guides/PersianCalendar.jsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import React from 'react';
+import SourcablePanel from '_shared/SourcablePanel';
+import { Typography } from 'material-ui';
+
+const CssOverrides = () => (
```
Comment (dmtrKovalenko) [inline docs/src/Routes/Guides/PersianCalendar.jsx]: Change the name please :)

File: docs/src/Routes/Guides/PersianCalendar.jsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      </p>
+    </Typography>
+
+    <Typography variant="body1" gutterBottom>
+      You will also need to install the <code> material-ui-pickers-jalali-utils </code>
+      package from npm.
+    </Typography>
+
+    <Typography variant="body1">
+      <code>npm install material-ui-pickers-jalali-utils </code>
```
Comment (dmtrKovalenko) [inline docs/src/Routes/Guides/PersianCalendar.jsx]: Maybe use `_shared/Code`

File: docs/src/layout/NavigationMenu.jsx [hunk 3]
```diff
@@ -43,6 +43,10 @@ const NavigationMenu = ({ classes }) => (
     <Link to="/guides/formats" className={classes.navLink}>
       <ListItem button> Global format customization </ListItem>
     </Link>
+
+    <Link to="/guides/persian" className={classes.navLink}>
```
Comment (dmtrKovalenko) [inline docs/src/layout/NavigationMenu.jsx]: I suppose it should be under `Localization` section :)

File: docs/src/Routes/Guides/PersianCalendar.jsx (hunk 1 above)
Comment (alitaheri) [reply docs/src/Routes/Guides/PersianCalendar.jsx]: ops 😱

File: docs/src/Routes/Guides/PersianCalendar.jsx (hunk 2 above)
Comment (alitaheri) [reply docs/src/Routes/Guides/PersianCalendar.jsx]: Ok, Thanks 👍

File: docs/src/layout/NavigationMenu.jsx (hunk 3 above)
Comment (alitaheri) [reply docs/src/layout/NavigationMenu.jsx]: I thought that was for using different libraries. That was my first thought anyway, I'll move it 👍

**Comment by alitaheri** [comment]
@dmtrKovalenko All done. Have another look, please.

**Comment by dmtrKovalenko** [comment]
Looks good :) Thank you

**Pull request merged**